File: python/memory_pool/llm.py
import re
from openai import AsyncOpenAI
from typing import AsyncIterator


def strip_thoughts(text: str) -> str:
    return re.sub(r"<thought>[\s\S]*?</thought>", "", text).strip()


def parse_json(text: str) -> dict | list:
    import json
    no_thoughts = re.sub(r"<thought>[\s\S]*?</thought>", "", text)
    no_fences = re.sub(r"```[\w]*\n?|\n?```", "", no_thoughts).strip()
    m = re.search(r"(\{[\s\S]*\}|\[[\s\S]*\])", no_fences)
    if not m:
        raise ValueError(f"No JSON found in: {no_fences[:100]}")
    sanitized = re.sub(r'\\([^"\\/bfnrtu])', r'\1', m.group(0))
    return json.loads(sanitized)


async def chat_complete(client: AsyncOpenAI, model: str, messages: list[dict], max_tokens: int = 512, temperature: float = 0.2) -> str:
    resp = await client.chat.completions.create(
        model=model,
        messages=messages,
        max_tokens=max_tokens,
        temperature=temperature,
    )
    return strip_thoughts(resp.choices[0].message.content or "")


async def stream_chat_complete(client: AsyncOpenAI, model: str, messages: list[dict], max_tokens: int = 2048, temperature: float = 0.7) -> AsyncIterator[str]:
    stream = await client.chat.completions.create(
        model=model,
        messages=messages,
        max_tokens=max_tokens,
        temperature=temperature,
        stream=True,
    )

    buf = ""
    in_thought = False

    async for chunk in stream:
        delta = chunk.choices[0].delta.content or ""
        if not delta:
            continue

        buf += delta
        output = ""

        while buf:
            if not in_thought:
                tag_start = buf.find("<thought>")
                if tag_start == -1:
                    lt = buf.rfind("<")
                    if lt != -1 and lt > len(buf) - 9:
                        output += buf[:lt]
                        buf = buf[lt:]
                    else:
                        output += buf
                        buf = ""
                    break
                output += buf[:tag_start]
                buf = buf[tag_start + 9:]
                in_thought = True
            else:
                tag_end = buf.find("</thought>")
                if tag_end == -1:
                    buf = buf[max(0, len(buf) - 10):]
                    break
                buf = buf[tag_end + 10:]
                in_thought = False

        if output:
            yield output

File: bun-npm/src/llm.ts
import OpenAI from 'openai'

export function stripThoughts(text: string): string {
  return text.replace(/<thought>[\s\S]*?<\/thought>/g, '').trim()
}

export function parseJSON(text: string): unknown {
  const noThoughts = text.replace(/<thought>[\s\S]*?<\/thought>/g, '')
  const noFences = noThoughts.replace(/```[\w]*\n?|\n?```/g, '').trim()
  const match = noFences.match(/(\{[\s\S]*\}|\[[\s\S]*\])/)
  if (!match) throw new SyntaxError(`No JSON found in: ${noFences.slice(0, 100)}`)
  const sanitized = match[0].replace(/\\([^"\\/bfnrtu])/g, '$1')
  return JSON.parse(sanitized)
}

export async function chatComplete(
  client: OpenAI,
  model: string,
  messages: OpenAI.Chat.ChatCompletionMessageParam[],
  maxTokens = 512,
  temperature = 0.2,
): Promise<string> {
  const res = await client.chat.completions.create({ model, messages, max_tokens: maxTokens, temperature })
  return stripThoughts(res.choices[0]?.message?.content ?? '')
}

export async function* streamChatComplete(
  client: OpenAI,
  model: string,
  messages: OpenAI.Chat.ChatCompletionMessageParam[],
  maxTokens = 2048,
  temperature = 0.7,
): AsyncGenerator<string> {
  const stream = await client.chat.completions.create({
    model, messages, max_tokens: maxTokens, temperature, stream: true,
  })

  let buf = '', inThought = false

  for await (const chunk of stream) {
    const delta = chunk.choices[0]?.delta?.content ?? ''
    if (!delta) continue

    buf += delta
    let output = ''

    while (buf.length > 0) {
      if (!inThought) {
        const tagStart = buf.indexOf('<thought>')
        if (tagStart === -1) {
          const ltIdx = buf.lastIndexOf('<')
          if (ltIdx !== -1 && ltIdx > buf.length - 9) {
            output += buf.slice(0, ltIdx); buf = buf.slice(ltIdx)
          } else { output += buf; buf = '' }
          break
        }
        output += buf.slice(0, tagStart); buf = buf.slice(tagStart + 9); inThought = true
      } else {
        const tagEnd = buf.indexOf('</thought>')
        if (tagEnd === -1) { buf = buf.slice(Math.max(0, buf.length - 10)); break }
        buf = buf.slice(tagEnd + 10); inThought = false
      }
    }

    if (output) yield output
  }
}
